portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -60,13 +68,52 @@ private static String convertToJsonWithGson(List<String> arraylist) {
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
       String personName = request.getParameter("person-name");
       String personMessage = request.getParameter("person-comment");
-      
+      String imageUrl = getUploadedFileUrl(request, "image");
       Entity commentEntity = new Entity("Comment");
       commentEntity.setProperty("name", personName);
       commentEntity.setProperty("messageContent", personMessage);
+      if (imageUrl != null) {
+          commentEntity.setProperty("imageUrl", imageUrl);
+      }
       DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
       datastore.put(commentEntity);
-
       response.sendRedirect("/comments.html");
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get("image");
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return null;
+    }
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+    String url = imagesService.getServingUrl(options);
+
+    // GCS's localhost preview is not actually on localhost,
+    // so make the URL relative to the current domain.
+    if(url.startsWith("http://localhost:8080/")){
```

Comment:
What is happening here?

Reply:
I am not very sure, I took it from the example on the tutorial. I can make some research to understand this 3 lines better.